### DISCUSSION TITLE
How do copilot suggestions happen?

### DISCUSSION THREAD

**Discussion by jjohnston-aims** [LazyVim/LazyVim#4589] [Q&A] [answered]
I look at [the docs](http://www.lazyvim.org/extras/coding/copilot) and see 

```
{
  "zbirenbaum/copilot.lua",
  cmd = "Copilot",
  build = ":Copilot auth",
  opts = {
    suggestion = { enabled = false },
    panel = { enabled = false },
    filetypes = {
      markdown = true,
      help = true,
    },
  },
}
```

So if panel and suggestions are both disabled, why do I see them in neovim when I am typing? I ask because `<CR>` will accept the suggestion from the drop down, not from ghost text and I'm trying to figure out what's going on.

**Comment by dpetka2001** [comment, marked as answer]
The Copilot suggestions happen from the `copilot` nvim-cmp source. From the [docs](https://github.com/zbirenbaum/copilot-cmp?tab=readme-ov-file#install) it suggests disabling these features in `copilot.lua`, that's why they're also disabled in LazyVim. The ghost text you're seeing is a `nvim-cmp` feature.

**Comment by utensil** [comment]
I also can't figure out how to accept the suggestion from the ghost text, tab doesn't work.

**Reply by aeciolevy** [reply]
Most of the time this works for me
https://github.com/LazyVim/LazyVim/discussions/4830#discussion-7507424

**Reply by utensil** [reply]
Thanks, I have switched from cmp to blink, and my setup to make copilot and cmp coexist better is to reserve tab for copilot, and use the following keys for blink completion

```lua
           keymap = {
                ["<C-q>"] = { "show", "hide" },
                ["<Up>"] = { "select_prev", "fallback" },
                ["<Down>"] = { "select_next", "fallback" },
                ["<CR>"] = { "accept", "fallback" },
                ["<Right>"] = { "accept", "fallback" },
                ["<Left>"] = { "hide" },
            },
```

So it works like this:

- with Github Copilot
    - suggestions show as ghost text
    - `Tab` to accept suggestions
    - `Shift+Tab` to use the normal Tab
- with `blink.cmp`
    - I've configure it to not triggered by default, in favor of Github Copilot
        - toggle by `Ctrl+q`
        - could also be triggerd by `down` in insert mode
    - up/down to cycle through suggestions, `Enter` to accept
    - the above is set up so all this key mapping are explicit, intuitive and won't interfere with `<Tab>` for Github Copilot
